src/util/parallel/data.ml

```diff
@@ -0,0 +1,242 @@
+        buckets = Atomic.make @@ Array.init size (fun _ -> Atomic.make None);
+      }
+
+    let to_list hm =
+      Array.fold_left (fun acc bucket ->
+          match Atomic.get bucket with
+          | None -> acc
+          | Some bucket -> acc @ Bucket.to_list bucket
+        ) [] (Atomic.get hm.buckets)
+
+    let to_seq hm =
+      let bucket_seq = Array.to_seq (Atomic.get hm.buckets) in
+      let non_atomic_bucket_seq = Seq.map Atomic.get bucket_seq in
+      let non_option_bucket_seq = Seq.filter (fun x -> x <> None) non_atomic_bucket_seq in
+      let non_atomic_bucket_seq = Seq.map (fun x -> match x with | Some x -> x | None -> failwith "This should not happen") non_option_bucket_seq in
+      Seq.flat_map Bucket.to_seq non_atomic_bucket_seq
+
+    let find_option hm key =
+      let hash = abs @@ H.hash key in
+      let bucket = Array.get (Atomic.get hm.buckets) (hash mod (Atomic.get hm.size)) in
+      match Atomic.get bucket with
+      | None -> None
+      | Some bucket -> Bucket.find_option bucket key
+
+    let find hm key =
+      match find_option hm key with
+      | None -> failwith "Key not found"
+      | Some value -> value
+
+    let mem hm key =
+      match find_option hm key with
+      | None -> false
+      | Some _ -> true
+
+
+    let rec find_create (hm : t) (key : H.t) =
+      let rec find_create_inner hm key hash buckets =
+        let bucket = Array.get buckets (hash mod (Atomic.get hm.size)) in
+        match Atomic.get bucket with
+        | None ->
+          let new_bucket = Bucket.create key in
+          let success = Atomic.compare_and_set bucket None (Some new_bucket) in
+          if success then (
+            (new_bucket.value, true) 
+          )
+          else find_create_inner hm key hash buckets 
+        | Some bucket -> 
+          let value, was_created = Bucket.find_create bucket key in
+          (value, was_created)
+      in
+      let current_generation = Atomic.get hm.resize_generation in
+      let hash = abs @@ H.hash key in
+      let value, was_created = find_create_inner hm key hash (Atomic.get hm.buckets) in
+      if (current_generation mod 2 == 0) && (Atomic.get hm.resize_generation == current_generation || not was_created) then (
+        if (Atomic.get hm.nr_elements >= Atomic.get hm.size * 2) then (
+          resize hm;
+        );
+        if (was_created) then Atomic.incr hm.nr_elements;
+        (value, was_created)
+      )
+      else (
+        while (Atomic.get hm.resize_generation == current_generation) do () (* spin *)
+        done;
+        find_create hm key)
+
+
+    and resize hm =
+      let current_generation = Atomic.get hm.resize_generation in
+      if ((current_generation mod 2 == 0) && Atomic.compare_and_set hm.resize_generation current_generation (current_generation+1)) then (
+
+        let old_size = Atomic.get hm.size in
+        let new_size = old_size * 2 in
+
+        let new_buckets = Array.init new_size (fun _ -> Atomic.make None) in
```

Comment:
Could this be `Array.make`? Or not because of physical equality? If not, we should leave a comment; otherwise, use it.

Reply:
We do, Atomics are references and having the same reference everywhere would of course not be helpful, adding the comment.